editoast/src/views/pagination.rs

```diff
@@ -12,11 +12,85 @@ use serde::Serialize;
 use thiserror::Error;
 use tracing::warn;
 use utoipa::IntoParams;
+use utoipa::ToSchema;
 
 use crate::error::Result;
 use crate::modelsv2::DbConnection;
+use crate::ListAndCount;
+use crate::Model;
+use crate::SelectionSettings;
 
-const DEFAULT_PAGE_SIZE: i64 = 25;
+editoast_common::schemas! {
+    PaginationStats,
+}
+
+const DEFAULT_PAGE_SIZE: u64 = 25;
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, ToSchema)]
+pub struct PaginationStats {
+    /// The total number of items
+    #[schema(minimum = 0)]
+    pub count: u64,
+
+    /// The number of items per page
+    #[schema(minimum = 1)]
+    pub page_size: u64,
+
+    /// The total number of pages
+    #[schema(minimum = 0)]
+    pub page_count: u64,
+
+    /// The current page number
+    #[schema(minimum = 1)]
+    pub current: u64,
+
+    /// The previous page number, if any
+    #[schema(required, minimum = 1)]
+    pub previous: Option<u64>,
+
+    /// The next page number, if any
+    #[schema(required, minimum = 1)]
+    pub next: Option<u64>,
+}
+
+impl PaginationStats {
+    pub fn new(current_page_count: u64, total_count: u64, page: u64, page_size: u64) -> Self {
+        assert!(page > 0);
+        assert!(page_size > 0);
+        assert!((page - 1) * page_size + current_page_count <= total_count);
+        let page_count = total_count.div_ceil(page_size);
+        let previous = (page > 1 && total_count > 0).then_some(page - 1);
+        let next = ((page - 1) * page_size + current_page_count < total_count).then_some(page + 1);
+        Self {
+            count: total_count,
+            page_size,
+            page_count,
+            current: page,
+            previous,
+            next,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+pub trait PaginatedList: ListAndCount + 'static {
+    /// Lists the models and compute [PaginationStats]
+    ///
+    /// See [ListAndCount::list_and_count] for more details.
+    ///
+    /// Panics if the limit or the offset of the `settings` are not set.
+    async fn list_paginated(
+        conn: &mut DbConnection,
+        settings: SelectionSettings<Self>,
+    ) -> Result<(Vec<Self>, PaginationStats)> {
+        let (page, page_size) = settings.as_pagination();
+        let (results, count) = Self::list_and_count(conn, settings).await?;
+        let stats = PaginationStats::new(results.len() as u64, count, page, page_size);
```

Comment:
The function already returns a `Result`. So you may use `u64::try_from(results.len())?` instead.

Reply:
I'm not sure `diesel` can return more than `i64::MAX` results so this conversion isn't supposed to fail. Would you rather `impl EditoastError for <u64 out of bounds error>` or `PaginatedListError::UnexpectedTooManyResults`?

Reply:
If we return this error to the client, maybe `PaginatedListError::UnexpectedTooManyResults` is better (less technical, more usable from the point of view of the consumer)?

Reply:
As discussed with @leovalais, let's not return such an error to the client. It should not even happen and if it does, it's an `editoast` problem so no reason to leak it to someone else. Not finding an easy way to do differently, we can leave it as it is, `panic!()` has the benefit of being visible for us developers (maybe later, when we have a more advanced management of our logs and alerts, we might be able to do differently).